### PR TITLE
Maintenance

### PR DESCRIPTION
**Pull request by pieterbreed**
The changes in this MR supports the work done for load-testing in gometro/bridge-telemetry.

## Summary

- updated dependencies including amongst others:
  - confluent/kafka
  - `com.clojure-goes-fast/clj-async-profiler` (flamegraphs)
  - `com.cnuernber/charred` (faster JSON)
- added type hints in a few places to save on reflection
- kafka related
  - counters for messages consumed and messages produced
  - `CaughtUpOnceNotifications` strategy. This is usedful for letting ktables finish loading once they've reached the end-offsets at the time of starting up, so they don't get stuck infinitely loading newer and newer data, without letting the rest of the system starting up.
  - Using charred JSON parsing library in deserializers/serializers
  - fine-tuning kafka utilities for deleting topics to prevent deleting internal and confluent topics

### CODE REVIEW

File: src/afrolabs/components/kafka.clj [hunk 1]
```diff
@@ -624,6 +644,66 @@
...
+        (let [topic-partition-assignment (.assignment ^KafkaConsumer consumer)
+
+              end-offsets
+              (into #{}
+                    (map (fn [[tp o]] {:topic (.topic ^TopicPartition tp)
+                                       :partition (.partition ^TopicPartition tp)
+                                       :offset o}))
+                    (.endOffsets ^KafkaConsumer consumer
+                                 topic-partition-assignment))]
+          (reset! end-offsets-at-start end-offsets)))
```
Comment (pieterbreed) [inline src/afrolabs/components/kafka.clj:679]: We store the end-offsets that are "current" right after the consumer is initialized.

File: src/afrolabs/components/kafka.clj [hunk 2]
```diff
@@ -624,6 +644,66 @@
...
+          [_ consumer _consumed-records]
+        (let [topic-partition-assignment (.assignment ^KafkaConsumer consumer)
+
+              current-offsets
+              (into #{}
+                    (map (fn [tp] {:topic (.topic ^TopicPartition tp)
+                                   :partition (.partition ^TopicPartition tp)
+                                   :offset (.position ^KafkaConsumer consumer
+                                                      ^TopicPartition tp)}))
+                    topic-partition-assignment)]
```
Comment (pieterbreed) [inline src/afrolabs/components/kafka.clj:692]: then after consuming, we take the current consumer offsets to compare them with the starting-end-offsets, to determine if we've surpassed that point for every partition yet.

File: src/afrolabs/components/kafka.clj [hunk 3]
```diff
@@ -1368,13 +1458,17 @@
 (s/def ::ktable-id (s/and string?
                           #(pos-int? (count %))))
 
+(s/def ::caught-up-once? boolean?)
 (s/def ::ktable-cfg (s/and ::clientless-consumer
-                           (s/keys :req-un [::ktable-id])))
+                           (s/keys :req-un [::ktable-id]
+                                   :opt-un [::caught-up-once?])))
```
Comment (pieterbreed) [inline src/afrolabs/components/kafka.clj:1464]: define `::caught-up-once?` flag for ktable configuration, which will optionally use the `CaughtUpOnceNotification` strategy for "completing" the initial load.

File: src/afrolabs/components/kafka.clj [hunk 4]
```diff
@@ -1410,7 +1506,18 @@
...
+          (loop []
+            (let [caught-up? (deref has-caught-up-once
+                                    5000
+                                    :timeout)]
+              (when (= :timeout caught-up?)
+                (log/info (str "KTable '" consumer-group-id "' loading initial data. Waited "
+                               (long (/ (- (System/currentTimeMillis) starting-millis)
+                                        1000))
+                               " seconds so far."))
+                (recur)))))
```
Comment (pieterbreed) [inline src/afrolabs/components/kafka.clj:1520]: When a caller is waiting for the ktable to load (by using `deref`/`@`) we produce a log statement every 5 seconds so the waiting is not invisible.

File: src/afrolabs/components/kafka/bytes_serdes.clj [hunk 5]
```diff
@@ -37,13 +38,22 @@
...
                               (alength ^bytes byte-data))))
+  ([this _ _ ^bytes byte-data]
+   (deser-deserialize-String-Headers-byte<> this nil byte-data)))
+
+(defn deser-deserialize-String-Headers-ByteBuffer
+  ([_ _ ^ByteBuffer byte-data]
+   (when byte-data
+     (let [result (make-array Byte/TYPE (.remaining byte-data))]
+       (.get byte-data ^bytes result)
+       result)))
```
Comment (pieterbreed) [inline src/afrolabs/components/kafka/bytes_serdes.clj:54]: The kafka deserializer interface unexpectedly gained a new overload on `deserialize`. This is a backwards compatible change for other JVM languages, but clojure has a bug where we are forced to implement this overload on `ByteBuffer`. More info here: https://pb.co.za/overloading-gen-class-methods-of-glorious-types-with-sublime-nation-of-clojure-for-make-great-benefit-of-interop

File: src/afrolabs/components/kafka/json_serdes.clj [hunk 6]
```diff
@@ -2,10 +2,20 @@
...
    :main false)
-  (:require [clojure.data.json :as json]
-            [taoensso.timbre :as log]
-            [clojure.string :as str])
-  (:import [org.apache.kafka.common.header Headers]))
+  (:require
+   ;; [clojure.data.json :as json]
+   [charred.api :as json]
+   [taoensso.timbre :as log]
+   [clojure.string :as str])
```
Comment (pieterbreed) [inline src/afrolabs/components/kafka/json_serdes.clj:9]: Using the `charred.api` library for loading JSON, which has better performance than `clojure.data.json` and with the same API.

File: src/afrolabs/config.clj [hunk 7]
```diff
@@ -75,9 +77,22 @@
...
 (defmethod aero/reader 'csv-array
   [_ _ value]
-  (first (csv/read-csv (java.io.StringReader. value))))
+  (interpret-string-as-csv-row value))
+
+(defmethod aero/reader 'long?
+  [_ _ value]
+  (when (and value
+             (seq value))
+    (Long/parseLong (str value))))
```
Comment (pieterbreed) [inline src/afrolabs/config.clj:92]: new aero (config file) reader tag for values when if present, must become long values, otherwise being nillable.

**Review by pieterbreed — COMMENTED** [review]
Author's notes

**Review by faf-gm — APPROVED** [review]

File: src/afrolabs/logging.clj [hunk 8]
```diff
@@ -21,14 +21,17 @@
...
-                 (atomic-println (str output
-                                      (when (seq context)
-                                        (str " :: [" context "]"))))))})
+  [& {:keys [colorized-output?]
+      :or   {colorized-output? false}}]
+  (cond-> {:enabled? true
+           :fn       (fn [data]
+                       (let [{:keys [output_ context]} data
+                             output                    (force output_)]
+                         (atomic-println (str output
```
[the remark this answers is not in the document]
Comment (pieterbreed) [reply src/afrolabs/logging.clj:30]: no, this hooks into the logging library. It does write to `*stdout*` as the normal/default logger does, but additionally prints the `*context*` value in the output.

File: src/afrolabs/aws/cloudwatch_logs.clj [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+    (csp/thread
+      (loop [to (csp/timeout batch-period-ms)]
+        ;; do uploading
+        (let [[msgs _] (swap-vals! enqued-log-events (constantly []))
+              batches (make-batches msgs)]
+          (doseq [batch batches]
+            (put-log-events client group stream batch)))
+        ;; wait for batch-period of time, or until we receive the signal to stop
+        (let [[_ ch] (csp/alts!! [stop-signal to])]
+          (when-not (= ch stop-signal)
```
[the remark this answers is not in the document]
Comment (pieterbreed) [reply src/afrolabs/aws/cloudwatch_logs.clj:166]: yes it will write what it has in the buffer and then not recur, and then terminate.

**Pull request merged**